db/db_impl.cc

```diff
@@ -5143,12 +5143,20 @@ Status DBImpl::DelayWrite(uint64_t num_bytes,
       if (write_options.no_slowdown) {
         return Status::Incomplete();
       }
-      mutex_.Unlock();
-      delayed = true;
-      TEST_SYNC_POINT("DBImpl::DelayWrite:Sleep");
-      // hopefully we don't have to sleep more than 2 billion microseconds
-      env_->SleepForMicroseconds(static_cast<int>(delay));
-      mutex_.Lock();
+      TEST_SYNC_POINT_CALLBACK("DBImpl::DelayWrite:Sleep", &delay);
+
+      // We will delay the write until the wall clock reach stall_end or
+      // we don't have any flushes or compactions running in the bg
+      uint64_t stall_end = sw.start_time() + delay;
+      while (bg_flush_scheduled_ > 0 || bg_compaction_scheduled_ > 0) {
+        uint64_t now = env_->NowMicros();
+        if (now >= stall_end) {
+          break;
+        }
+
+        delayed = true;
+        bg_cv_.TimedWait(static_cast<int>(stall_end - now));
```

Comment:
@siying, I think you mean there is a problem with NowMicros() not TimedWait(), correct ?
NowMicros() is the one that will be affected by system time change, correct ?